### ISSUE TITLE
Confirm Config is Never Overwritten

### ISSUE THREAD

**Issue by F483** [Storj/dataserv-client#118]
*heunland* reported her config was overwritten when the build path was changed.

**Comment by littleskunk** [comment]
Maybe it was the config migration?

**Comment by F483** [comment]
It should never change the payout address and the only actual migration that changes anything but the version number is https://github.com/Storj/dataserv-client/blob/master/dataserv_client/control/config.py#L71 which converts the master_secret to a hwif (preserving the auth wif).

**Comment by super3** [comment]
@littleskunk can you see if you can replicate this.

**Comment by littleskunk** [comment]
I asked heunland what she did. She is using windows 7 and the default config path. She was building on a fat32 drive with different setups. The final build was on Drive C again and it overwrites the default config.

Now here is what i tryed:
Create a 2.0.1 config and start 2.1.3 client with config, config --set_payout_address, register and build. I tryed different store_path as well. Everything without luck. I always had my payout address.
Invalid old or new configs will result in a invalid config message without overwriting.

I see only one way. Change windows user and you will have a new default config path. Thats why i vote for a bugfix.
Possible solution:
1) At the moment config, register, build an farm will create a "useless" default config if it does not exists. A missing config exception would be better. Does anybody need the "useless" default config?
2) dataserv-client config should not write a default config as well. What do you think about dataserv-client config --autogenerate ? Only this command should write a "useless" default config.

**Comment by heunland** [comment]
skunk the config that was overwritten was one I had made using my own crowdsale address. It was overwritten by a default config but the original was one I had made to farm on my own addy.

**Comment by littleskunk** [comment]
I created default config and payout config but i was not able to overwrite them in any way. Bugfix should work for that as well.

@super3 @F483 do you agree with the bugfix?

**Comment by super3** [comment]
I like option number 1 best. Can you put in a pull request for that @littleskunk?

**Comment by littleskunk** [comment]
Pull request ready. At the moment unit test are broken but i did my best to keep my part clear.

**Comment by F483** [comment]
I'm very sorry I'm late to the party, somehow this issue advanced without me noticing and now [this](https://github.com/Storj/dataserv-client/pull/141/files) pull request exists.

I see several problems with what happened.

1. The original problem that @heunland had was never reproduced as far as I can tell. This is 'fix' is trying to address a bug we don't exactly know. This is called poking in the dark hoping to fix something, but maybe just breaking something else instead.

2. Instead of solving the problem @heunland had we now changed the way the config works. Now the user is forced to configure something to get a config before they can farm. Additionally this may introduce new bugs we don’t know of yet. I originally wrote it so that the user would not have to configure anything by default. I do not think this is a good change.

Now that we have this patch, the question is what do we do. Request for comment @super3 @littleskunk @heunland 

As such the changes to control.config are minimal and I wanted to move that code to storjcore anyway so touching it again wont cause to much trouble. The question is do we want to force the user to configure beforehand or not?

If not I suggest we keep the original implementation and allow the user to configure things as they go. If so I would suggest instead having an initial configuration procedure (like creating ssl keys).

I personally am for creating a sane default config and letting the user configure as they go. They will likely be changing and tweaking things all the time as more options get added in the future. It will be essential for farmers to be flexible and adjust there config (discs, size, rates, etc ...) quickly and even automatically via scripts. I think a set initialization procedure will quickly be inadequate.

**Comment by littleskunk** [comment]
I agree that the fix is not for @heunland but i think we need it anyway. If the config file is missing (wrong user account or missing --config_path) farm and register will silently generate a useless config file with useless payout addr.

2) not implemented. "dataserv-client" config will work like befor.

**Comment by F483** [comment]
I do not think it is a useless config, the payout address can be any bitcoin address and generating a wallet for the user to import elsewhere is a good thing.

Generating a config silently is not a bad thing, this is commonly done by most apps and as long as its in an expected location like "~/.storj" it is no problem.

**Comment by heunland** [comment]
ok I was asked to comment on this issue. I agree with Fabe, I am not sure what exactly will happen with this fix or what implications it may have ... all I know is that I was surprised that my previously created config was replaced with a default one when I tried to farm without specifying any config argument. What I can say now however is that I just updated this machine to v2.1.4 and just ran the farm command again without doing the config step first, and this time it correctly recognized my previous config.json and started farming with no errors.
